server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java

```diff
@@ -177,11 +193,14 @@ public void reanalyseOpenedPrograms() throws InterruptedException {
             .stream().filter(d -> !analysisService.isCopybook(d.getUri(), d.getText())).collect(Collectors.toList());
     boolean analysisInProgress;
     do {
-      analysisInProgress = openDocuments.stream()
+        analysisInProgress = openDocuments.stream()
               .filter(model -> model.getLastAnalysisResult() != null)
-              .map(model -> makeId(model.getUri(), analysisResultsRevisions.get(model.getUri())))
-              .filter(analysisResults::containsKey)
-              .anyMatch(id -> !analysisResults.get(id).isDone());
+              .filter(model -> {
+                String id = makeId(model.getUri(), analysisResultsRevisions.get(model.getUri()));
+                return analysisResults.containsKey(id);
+              })
+              .anyMatch(model -> !model.isLastAnalysisCancelled());
```

Comment:
idea is to consider a analysis cancelled when the callable reaches an exception where we update this flag in the documentModel. In case the computation is not interrupted, this future would be removed from the analysisResults map and would not be considered as ongoing computation. 
But we continue to check if cancel still works in a do-while loop.

Reply:
or just shutdown Executor Service, but this sounds resource extensive operation.